src/v/utils/named_type.h

```diff
@@ -33,24 +33,14 @@ class base_named_type<T, Tag, std::true_type> {
     base_named_type& operator=(base_named_type&& o) noexcept = default;
```

Comment:
> default when possible.
> 
> <=> can't be easily defaulted, by the rules of cpp. the solution is to
> manually implement it if the type does not provide it

I think I'm missing the part where the missing <=> is allowing the surprising comparisons. Or is it that our previous implementations of the comparators were too lax, and the new implementation with <=> is less lax?

Reply:
yes it's the fact that previous definitions allowed implicit conversions to take place.
i have to find the reference in the standard, but the idea is that before there was a path to convert a base_named_type to its T, and this conversion was available to use in operator< of another base_named_type with compatible T. this form will not take part in the conversion (but it's an accidental discovery, i was actually writing something else :D )

you can test this with the static asserts in named_type_tests.cc, 

```cpp

static_assert(
  !std::equality_comparable_with<
    named_type<int, struct tag_0>,
    named_type<int, struct tag_1>>,
  "arithmetic named_types can't be compared directly");
static_assert(
  !std::equality_comparable_with<
    named_type<ss::sstring, struct tag_a>,
    named_type<ss::sstring, struct tag_b>>,
  "non-arithmetic named_types can't be compared directly");
 ```
 
 before this pr, this will not compile